Make stat cancellable in case of stuck paths

saketjajoo opened the pull request:
In cases when the underlying volume is stuck or not responding, the calls made to os.Stat while mounting such volumes gets stuck which makes containerd unusable. This PR executes stat via exec.CommandContext to which a timed context is passed which so that an answer is returned back eventually even when the underlying volume does not respond.

Fixes: https://github.com/containerd/containerd/issues/10828

chrishenzie commented on internal/cri/opts/spec_linux_opts.go:
```diff
@@ -106,11 +107,65 @@ func withMounts(osi osinterface.OS, config *runtime.ContainerConfig, extra []*ru
+
+				newInflight.Cond.Broadcast()
+				mu.Unlock()
+			}
```
It seems to me like we're leaking implementation details of `osManager` here. I would expect this method would call out to `osManager.Stat()`, and then the `osManager` will internally figure out whether to issue a new `stat` or bundle this request with one already in-flight.

chrishenzie commented on internal/cri/server/container_create_linux_test.go:
```diff
@@ -601,6 +606,135 @@ func TestMountPropagation(t *testing.T) {
+		// 	t.Logf("FUSE process exited with error: %v\n", fuseCmdErr)
+		// 	t.Logf("FUSE process stdout:\n%s\n", outBuffer.String())
+		// 	t.Logf("FUSE process stderr:\n%s\n", errBuffer.String())
+		// }
```
Are these lines needed?

saketjajoo replied:
These are not really needed. It's just for debugging if something goes wrong with FUSE itself while testing. I've removed it in the latest commit.

chrishenzie commented on internal/cri/server/podsandbox/controller.go:
```diff
@@ -82,7 +82,8 @@ func init() {
 				client:         client,
 				config:         runtimeService.Config(),
-				os:             osinterface.RealOS{},
+				os:             &osinterface.RealOS{},
```
Why do we need to change this to a pointer type?

chrishenzie commented on pkg/os/os.go:
```diff
@@ -91,3 +93,52 @@ func (RealOS) WriteFile(filename string, data []byte, perm os.FileMode) error {
+	Err error
+}
+
+type Manager interface {
```
`Manager` seems a bit general, are there other use cases besides `stat` for this type? If not, I'd prefer to keep this type more focused and we can refactor later if that changes (e.g. Change this to `StatManager` or something of that sort).

chrishenzie commented on pkg/os/os.go:
```diff
@@ -91,3 +93,52 @@ func (RealOS) WriteFile(filename string, data []byte, perm os.FileMode) error {
+	NewInflightStat() *InflightStat
+}
+
+type RealManager struct {
```
I don't think this type should be exported (e.g. `realManager`).

chrishenzie commented on pkg/os/os.go:
```diff
@@ -91,3 +93,52 @@ func (RealOS) WriteFile(filename string, data []byte, perm os.FileMode) error {
+
+// InflightStat represents a Stat call that is in progress for a given path.
+// It is exported so that consumers of the OS interface can use it to coordinate.
+type InflightStat struct {
```
This seems like an implementation detail we wouldn't want to leak to callers (e.g. make it `inflightStat`).

saketjajoo replied:
Done.

chrishenzie commented on pkg/os/os.go:
```diff
@@ -91,3 +93,52 @@ func (RealOS) WriteFile(filename string, data []byte, perm os.FileMode) error {
+type Manager interface {
+	StatTimeout() time.Duration
+	InflightStats() map[string]*InflightStat
+	InflightStatsMutex() *sync.Mutex
```
Should we be leaking these implementation details to callers?

chrishenzie commented on pkg/os/os.go:
```diff
@@ -91,3 +93,52 @@ func (RealOS) WriteFile(filename string, data []byte, perm os.FileMode) error {
+	Err error
+}
+
+type Manager interface {
```
I think this interface needs some attention. At a high-level, I'd imagine this type would be a thin wrapper on top of the existing `OS` interface, but handling the details of in-flight requests. Working with it could look like:
```go
os := osinterface.RealOS{}
statManager := osinterface.StatManager(os)
result := statManager.Stat("foo")
```

Inside of `statManager` would be this `OS` interface type, which we'd call out to once we're ready to actually `stat` a path:
```go
type statManager struct {
  os OS
}

func (m *statManager) Stat(name string) (os.FileInfo, error) {
  // all of the internal details for tracking in-flight requests, etc.
  if newRequest() {
    m.os.Stat(name)
  }
}
```

saketjajoo replied:
Done.

chrishenzie commented on pkg/os/testing/fake_os.go:
```diff
@@ -252,3 +253,35 @@ func (f *FakeOS) Hostname() (string, error) {
 	return "", nil
 }
+
+type FakeOSManager struct {
```
I think this type isn't necessary if the `statManager` would be a wrapper around the `OS` interface.

chrishenzie commented on internal/cri/server/podsandbox/controller_test.go:
```diff
@@ -45,9 +45,10 @@ var testConfig = criconfig.Config{
-		store:  NewStore(),
+		config:    testConfig,
+		os:        ostesting.NewFakeOS(),
+		osManager: ostesting.NewFakeOSManager(),
```
I think this "fake" type shouldn't be necessary if the `statManager` wraps the `OS` interface (which could be either a real or fake implementation):
```go
os := ostesting.NewFakeOS()
return &Controller{
  ...
  os: os,
  // The same type you would use in the actual controller.
  osManager: osinterface.NewStatManager(os),
}
```

saketjajoo commented on integration/failpoint/cmd/fake-FUSE/main.go:
```diff
@@ -0,0 +1,230 @@
+var fuseDir = flag.String("fuse-dir", "fakeFUSEDir", "dir to be used by the FUSE process to store everything.")
+
+// FUSE opts
+var hangDuringStat = flag.Bool("hang-during-stat", false, "if set to true, hang-during-stat will hang the Stat call made to FUSE's root dir.")
```
Yes, both are FUSE opts. However, the one above (`fuseDir`) is the opt for where the FUSE FS should be set up / initialized and the the one below (`hangDuringStat`) is how the FUSE FS should behave. Since both represent different set of opts for FUSE, I separated them. Let me know if this rationale sounds good.

saketjajoo commented on contrib/Dockerfile.test.d/critest.sh:
```diff
@@ -1,5 +1,21 @@
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
```
This was autogenerated while following one of the steps in [general contributing guidelines](https://github.com/containerd/project/blob/main/CONTRIBUTING.md#applying-license-header-to-new-files).

chrishenzie replied:
Revisiting this -- I would prefer we only include the `ltag` changes on the new files you introduced, and we can clean up any missing headers in a separate PR. This would help focus the scope of this PR to just what needs to change to fix this `stat` issue.

saketjajoo commented:
/retest

saketjajoo commented:
/cc mxpv

saketjajoo commented:
There's some flake in the CI environments. For example, earlier (see [slack message](https://cloud-native.slack.com/archives/CGEQHPYF4/p1752684865869259)), the `container_create_linux_test/TestPrivilegedDevices` test was failing with an error "_failed to generate spec: error stating device path: stat /dev/disk/by-loop-ref: no such file or directory_". However, upon adding a [dummy commit](https://github.com/containerd/containerd/pull/12023/commits/efb0ad707f0c0419f8f6d2897e55b24acd9a2be0), the test passed. Then, upon rebasing the branch onto `containerd/main`, the test is [passing in 1 CI](https://github.com/containerd/containerd/actions/runs/16432104445/job/46435445343?pr=12023) and failing in [another](https://github.com/containerd/containerd/actions/runs/16432104445/job/46435445429?pr=12023).

saketjajoo commented:
/retest

chrishenzie commented on pkg/os/os.go:
```diff
@@ -91,3 +95,89 @@ func (RealOS) WriteFile(filename string, data []byte, perm os.FileMode) error {
+		select {
+		case res := <-resultChan: // Stat completed
+			statError = res.Err
+		case <-statCtx.Done(): // timeout / mountpath could be stuck.
```
When `statCtx.Done()` returns, it seems like it could be for one of two reasons:
- The timeout was exceeded (`context.DeadlineExceeded`)
- Or maybe the context was cancelled for some other reason

Would it be worth differentiating these cases?
```go
if errors.Is(statCtx.Err(), context.DeadlineExceeded) {
    statError = fmt.Errorf("stat call for mountpath %q timed out after %v", path, sm.statTimeout)
} else {
    statError = fmt.Errorf("stat call for mountpath %q canceled: %w", path, statCtx.Err())
}
```

saketjajoo replied:
Makes sense. Done.

saketjajoo commented on pkg/os/os.go:
```diff
@@ -91,3 +95,89 @@ func (RealOS) WriteFile(filename string, data []byte, perm os.FileMode) error {
+func NewStatManager(os OS) *StatManager {
+	return &StatManager{
+		os:                 os,
+		statTimeout:        5 * time.Second,
```
For some systems 5s could be quite sufficient, but for others (e.g. slow network where the FS is remote [NFS], or the disk is inherently slow, etc.), 5s could be insufficient.

* Making it configurable can help but it may come at a cost of restarting containerd to pick up the new timeout (say if we choose to use the config file to specify this timeout).
* Further, there could be several filesystems on a machine, for some of which 5s (or the timeout specified in the configuration) could be sufficient and for others, not. Making it configurable may not help in this case unless the user configures this timeout to be a very large value which can handle all filesystems on the machine. I assume that an end user may not know how long a FS might take to respond to `Stat`, thus making this selection of value non-trivial.

Perhaps, we can discuss more on this and the [other comment](https://github.com/containerd/containerd/pull/12023#issuecomment-3128263995) in the community meeting to get other's feedback on this.

_(keeping this thread open for comments and discussions)_

saketjajoo commented:
/retest

saketjajoo commented:
/retest

saketjajoo commented:
/retest

saketjajoo commented:
/cc mikebrow

saketjajoo commented:
/retest

saketjajoo commented:
/retest

saketjajoo commented:
/retest

mxpv commented on pkg/os/os.go:
```diff
@@ -91,3 +96,93 @@ func (RealOS) WriteFile(filename string, data []byte, perm os.FileMode) error {
+
+// StatManager wraps an OS interface to provide managed access to Stat calls,
+// specifically to handle in-flight requests gracefully.
+type StatManager struct {
```
@chrishenzie any specific reason we want to keep it typeless? I'd rather keep one interface for OS interactions, so we don't have to pass 2 interfaces everywhere (`stat` is pretty popular call).

mxpv commented on pkg/os/os.go:
```diff
@@ -91,3 +96,93 @@ func (RealOS) WriteFile(filename string, data []byte, perm os.FileMode) error {
+		resultChan := make(chan StatResult, 1)
+		go func() {
+			_, err := sm.os.Stat(path)
+			resultChan <- StatResult{Err: err}
```
Does `sm.os.Stat` ever exits? Or we end up with leaked goroutine forever here?

mxpv commented on pkg/os/os.go:
```diff
@@ -91,3 +96,93 @@ func (RealOS) WriteFile(filename string, data []byte, perm os.FileMode) error {
+}
+
+// Stat performs a stat on the given name, managing concurrent requests to the same path.
+func (sm *StatManager) Stat(ctx context.Context, path string) error {
```
Is the goal just to verify `path` exists?
If so, there is much easier solution:

```go
  func (os OSInterface) Exists(path string, timeout time.Duration) error {                                                                                                                              
      ctx, cancel := context.WithTimeout(context.Background(), timeout)                                                                                                                         
      defer cancel()                                                                                                                                                                            
                                                                                                                                                                                                
      cmd := exec.CommandContext(ctx, "stat", path)                                                                                                                                             
      return cmd.Run() // Kill after timeout, no goroutines leaks, trivial implementation
  }
```

mxpv reviewed (changes requested):
I have some concerns about both the general approach taken here and a few specific implementation details

samuelkarp reviewed (commented):
Can you separate out the introduction of `fake-fuse` into a separate commit?  That's worth keeping even if we change the implementation to simplify as @mxpv suggested.

saketjajoo commented:
/retest

mikebrow commented on pkg/os/os.go:
```diff
@@ -59,6 +64,35 @@ func (RealOS) Stat(name string) (os.FileInfo, error) {
+// Exists will execute "stat" with a context to verify if the given path exists.
+// If the context does not have a deadline set, then a default 5 sec timeout
+// will be used while calling stat.
+func (RealOS) Exists(ctx context.Context, path string) error {
```
sounds like an opportunity then to do some housekeeping... thanks so much for the log

samuelkarp commented on internal/cri/server/helpers_test.go:
```diff
@@ -40,6 +41,11 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+func TestEmpty(t *testing.T) {
```
Which test needs this?  It should be in the file that needs it.

samuelkarp commented on integration/failpoint/cmd/fake-fuse/main.go:
```diff
@@ -0,0 +1,274 @@
+	if err := os.MkdirAll(fuseDirectory, os.ModeDir); err != nil {
+		log.G(ctx).Errorf("failed to create a fuse directory (%v): %v", fuseDirectory, err)
+	}
+	// If the user has provided the fuseDir, then they are responsible to manage it.
```
Since we have exactly one use-case for this fuse implementation right now, we should match that use-case. We should either have the test manage the directory or this program, not conditionally switch.

samuelkarp commented on integration/failpoint/cmd/fake-fuse/main.go:
```diff
@@ -0,0 +1,274 @@
+	if *hangDuringStat {
+		log.G(ctx).Info("setting up FUSE to hang during Stat")
+		var statHangWait fifosync.Waiter
+		var err error
```
style nit: we usually group if there are multiple consecutive `var`s:

```go
var (
	statHangWait fifosync.Waiter
	err          error
)
```

samuelkarp commented on integration/failpoint/cmd/fake-fuse/main.go:
```diff
@@ -0,0 +1,274 @@
+	log.SetLevel(*logLevel)
+	flag.Parse()
+	flag.VisitAll(func(f *flag.Flag) {
+		log.G(ctx).Debugf("fakeFUSE flag %s: %s\n", f.Name, f.Value)
```
We generally prefer structured logs rather than printf style:

```go
log.G(ctx).WithField("flag", f.Name).WithField("value", f.Value).Debug("fakeFUSE flag")
```

saketjajoo committed:
Fix the typo in the test name

Signed-off-by: Saket Jajoo <saketjajoo@google.com>

saketjajoo committed:
Implement FUSE.

FUSE is used to test the stuck Stat issue
(https://github.com/containerd/containerd/issues/10828).

Signed-off-by: Saket Jajoo <saketjajoo@google.com>

saketjajoo committed:
Add a unit test for testing issue/10828.

This unit test should fail because stat does not return within the given
timeout while building the container spec. The test uses FUSE to mock a
filesystem which is stuck and not responding to any stat calls.

Tests: https://github.com/containerd/containerd/issues/10828
Signed-off-by: Saket Jajoo <saketjajoo@google.com>

saketjajoo committed:
Make stat cancellable in case of stuck paths

In cases when the underlying volume is stuck or not responding, the
calls made to os.Stat while mounting such volumes gets stuck which makes
containerd unusable. This PR executes stat via exec.CommandContext to
which a timed context is passed which so that an answer is returned back
eventually even when the underlying volume does not respond.

Fixes: https://github.com/containerd/containerd/issues/10828
Signed-off-by: Saket Jajoo <saketjajoo@google.com>

saketjajoo commented on internal/cri/server/container_create_test.go:
```diff
@@ -70,7 +69,7 @@ func TestGeneralContainerSpec(t *testing.T) {
 	testSandboxID := "sandbox-id"
 	testContainerName := "container-name"
-	spec, err := c.buildContainerSpec(currentPlatform, testID, testSandboxID, testPid, "", testContainerName, testImageName, containerConfig, sandboxConfig, imageConfig, nil, ociRuntime, nil)
+	spec, err := c.buildContainerSpec(t.Context(), currentPlatform, testID, testSandboxID, testPid, "", testContainerName, testImageName, containerConfig, sandboxConfig, imageConfig, nil, ociRuntime, nil)
```
I've tried to split and reorder the commits. Now it looks like (from bottom-most to top-most commit):
1. Fix the typo in test name.
2. Implement FUSE (just implementation, no usage).
3. Adding a unit test for the issue (this test is expected to fail).
4. Fixing the issue and modifying related codepaths.

prakleumas commented on internal/cri/server/container_create_linux_test.go:
```diff
@@ -647,6 +652,139 @@ func TestMountPropagation(t *testing.T) {
+		{
+			ContainerPath: "container-path-3",
+			HostPath:      workingDir,
+		},
```
This also introduces the opportunity for a goroutine leak on a timeout. It is generally safer to protect this with a buffered channel.

```suggestion
	var err3Ch = make(chan error, 1)
```

prakleumas commented on pkg/os/os.go:
```diff
@@ -30,7 +34,7 @@ import (
 	MkdirAll(path string, perm os.FileMode) error
 	RemoveAll(path string) error
-	Stat(name string) (os.FileInfo, error)
+	Stat(ctx context.Context, name string) error
```
Modifying the exported `OS` interface by altering the `Stat` method signature—specifically adding a `context.Context` parameter and removing the `os.FileInfo` return type—introduces a breaking API change for downstream consumers. Under strict semantic versioning protocols, breaking changes to exported contracts are typically reserved for major version boundaries. 

To maintain backward compatibility, it is generally safer to retain the existing `Stat` method and introduce a distinct method for the context-aware operation.

```suggestion
	Stat(name string) (os.FileInfo, error)
	StatContext(ctx context.Context, name string) error
```

prakleumas commented on integration/failpoint/cmd/fake-fuse/main.go:
```diff
@@ -0,0 +1,267 @@
+	}()
+
+	err := fuseStruct.fuseCloseWait.WaitCtx(ctx)
+	if err != nil {
```
nit: scope err closer to use

```suggestion
	if err := fuseStruct.fuseCloseWait.WaitCtx(ctx); err != nil {
```

prakleumas commented on integration/failpoint/cmd/fake-fuse/main.go:
```diff
@@ -0,0 +1,267 @@
+	err := fs.fuseReadyTrigger.Trigger()
+	if err != nil {
+		log.G(fs.ctx).WithError(err).Fatalf("failed to trigger FUSE's readiness")
+	}
```
nit: scope err closer to use

```suggestion
	if err := fs.fuseReadyTrigger.Trigger(); err != nil {
```

prakleumas commented on integration/failpoint/cmd/fake-fuse/main.go:
```diff
@@ -0,0 +1,267 @@
+
+	fuseStruct := createFUSEFS(ctx)
+	log.G(ctx).Debugf("fuse struct: %v", fuseStruct)
+
```
It is generally safer to specify explicit directory permissions (such as `0755` or `0700`) rather than `os.ModeDir`. Passing `os.ModeDir` evaluates to a permission mask of `0000` (as it only sets the directory type bit), which introduces the opportunity for permission denied errors unless the process is always running as root.

```suggestion
	if err := os.MkdirAll(fuseDirectory, 0755); err != nil {
```

prakleumas commented on integration/failpoint/cmd/fake-fuse/main.go:
```diff
@@ -0,0 +1,267 @@
+	}
+
+	if *fuseDir == "" {
+		if err := os.RemoveAll(*fuseDir); err != nil {
```
This condition prevents the cleanup of the FUSE directory. The `main` function explicitly exits if `*fuseDir == ""`, guaranteeing that this block is never reached. To ensure correct teardown mechanics, this should evaluate if the directory string is populated.

prakleumas commented on internal/cri/server/container_create_linux_test.go:
```diff
@@ -647,6 +652,139 @@ func TestMountPropagation(t *testing.T) {
+	statHangResumeFIFOFile := filepath.Join(testDir, fuseFIFODirectoryName, statHangFIFOFileName)
+
+	// This path comes from the script/setup/install-failpoint-binaries file.
+	fuseRunCmd := filepath.Join("/usr/local/bin", "fake-fuse")
```
This is expedient, but not a great practice as now our test depends on the environment in which it runs (expecting `fake-fuse` to be installed in `/usr/local/bin`). I don't really think it's worth revisiting right now, but for the future I think the re-exec pattern (defining the program we want to execute as part of the test code) is probably more preferred to maintain hermetic test environments.

prakleumas commented on integration/failpoint/cmd/fake-fuse/main.go:
```diff
@@ -0,0 +1,267 @@
+				log.G(ctx).WithError(err).Error("failed to wait in stat")
+			}
+			return err
+		}
```
Spawning a separate goroutine to call `Wait()` while simultaneously calling `WaitCtx(ctx)` (which internally calls `Wait()`) introduces a race condition. Two concurrent readers on the same FIFO might cause one to consume the trigger while the other blocks indefinitely. Since `errCh` is unused, it is generally safer to remove this redundant goroutine and rely entirely on `WaitCtx`. Scoping the error variable is also a nice cleanup.

```suggestion
		attrFn = func(ctx context.Context, a *fuse.Attr) error {
			log.G(ctx).WithField("stat_hang_wait", statHangWait.Name()).Info("waiting during stat to repro issue 10828")

			if err := statHangWait.WaitCtx(ctx); err != nil {
				log.G(ctx).WithError(err).Error("failed to wait in stat")
				return err
			}
			return nil
		}
```

prakleumas reviewed (commented):
<sup>beep boop</sup>
Hello! I am @samuelkarp's friendly code review robot 🤖! Sam asked me to take a look and post my comments directly on GitHub. 

Please treat my comments as suggestions and advisory, not mandates. Since I am an LLM and my output is used here directly, you should evaluate carefully and see if my suggestions make sense. Feel free to directly resolve any you disagree with.

<sup>This is _not_ an automated action. This bot does not read or reply to comments.</sup>

saketjajoo commented:
/retest

saketjajoo commented on internal/cri/server/container_create_linux_test.go:
```diff
@@ -647,6 +652,139 @@ func TestMountPropagation(t *testing.T) {
+	fuseShutdownFIFOFile := filepath.Join(testDir, fuseFIFODirectoryName, fuseShutdownFIFOFileName)
+	statHangResumeFIFOFile := filepath.Join(testDir, fuseFIFODirectoryName, statHangFIFOFileName)
+
+	// This path comes from the script/setup/install-failpoint-binaries file.
+	fuseRunCmd := filepath.Join("/usr/local/bin", "fake-fuse")
+	fuseCmd := exec.CommandContext(context.Background(), fuseRunCmd, "--fuse-dir", testDir, "--hang-during-stat")
+	err := fuseCmd.Start()
+	require.NoErrorf(t, err, "failed to start the FUSE process via %q", fuseRunCmd)
```
It is okay to assume this as the test runs as a part of GH Actions workflow which [installs the failpoint binaries](https://github.com/containerd/containerd/blob/8b396c768d470226f120605ee5ad38d1a764bf81/.github/workflows/ci.yml#L400) before running the tests.

saketjajoo commented on pkg/os/os.go:
```diff
@@ -54,9 +58,35 @@ func (RealOS) RemoveAll(path string) error {
+		ctx, cancel = context.WithTimeout(ctx, 5*time.Second)
+		defer cancel()
+	}
+	cmd := exec.CommandContext(ctx, "stat", name)
+	err := cmd.Run()
+	if err != nil {
+		if ctx.Err() != nil { // Context expired
+			return ctx.Err()
+		}
```
WaitDelay would send a SIGKILL once the context is done. A stat on a unresponsive NFS/CIFS would be a D-state process which ignores SIGKILLs. So, using  `cmd.WaitDelay` may not help.

saketjajoo commented on pkg/os/os.go:
```diff
@@ -54,9 +58,35 @@ func (RealOS) RemoveAll(path string) error {
+		ctx, cancel = context.WithTimeout(ctx, 5*time.Second)
+		defer cancel()
+	}
+	cmd := exec.CommandContext(ctx, "stat", name)
```
In `pks/os.go`, I've removed `Stat` completely. I've added the `Stat`'s implementation in `pkg/os_unix.go` and `pkg/os_windows.go` to be platform dependent.

saketjajoo commented on pkg/os/os.go:
```diff
@@ -54,9 +58,35 @@ func (RealOS) RemoveAll(path string) error {
+			if exitErr.ExitCode() == 1 {
+				return os.ErrNotExist
+			}
+			return fmt.Errorf("stat failed with %v: %w", exitErr.ExitCode(), err)
```
Updated the `Stat` codepath to not parse the error string.

saketjajoo commented on pkg/os/os.go:
```diff
@@ -54,9 +58,35 @@ func (RealOS) RemoveAll(path string) error {
+		// In some environments, if a path is restricted, it might also return 1.
+		if exitErr, ok := err.(*exec.ExitError); ok {
+			if exitErr.ExitCode() == 1 {
+				return os.ErrNotExist
```
I've updated the `Stat` function to not parse the error string and make assumptions based on that. Instead, what it does now is - it runs `exec.CommandContext(ctx, "stat", path)` (with a timeout) to verify if the underlying path is responsive or not. If it is, then it calls `os.Stat(path)` and returns the error back.

saketjajoo commented:
/retest
